### PR TITLE
Add fileid preview endpoint

### LINKED ISSUE #7700: Do not regenerate the preview on a rename
* having a file with a preview
* rename the file
* actual: preview is hidden behind a spinner and then downloaded again
* expected: preview just stays

The reason for this is, that the preview is actually requested based on the filename and not on the fileid (it is nevertheless stored under the fileid and thus not re-generated).

Before and after rename URL:

```
http://192.168.99.100/server/core/preview.png?file=%2FBildschirmfoto+2017-11-30+um+23.09.2822.png&c=64e26567fd27cdb34123247436d80d29&x=32&y=32&forceIcon=0
http://192.168.99.100/server/core/preview.png?file=%2FBildschirmfoto+2017-11-30+um+23.09.28224.png&c=64e26567fd27cdb34123247436d80d29&x=32&y=32&forceIcon=0
```

cc @rullzer @nextcloud/javascript

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #7700 

File paths can change. The fileid is stable.

To observer watch your network console:
1. Upload an image
2. Open sidebar for that image
3. Now rename the file
4. Observe no new network request but the preview is still there

:balloon:

### CODE REVIEW

**Comment by rullzer** [comment]
@MorrisJobke this is what you had in mind right?

File: core/routes.php [hunk 1]
```diff
@@ -57,8 +57,8 @@
 		['name' => 'TwoFactorChallenge#showChallenge', 'url' => '/login/challenge/{challengeProviderId}', 'verb' => 'GET'],
 		['name' => 'TwoFactorChallenge#solveChallenge', 'url' => '/login/challenge/{challengeProviderId}', 'verb' => 'POST'],
 		['name' => 'OCJS#getConfig', 'url' => '/core/js/oc.js', 'verb' => 'GET'],
-		['name' => 'Preview#getPreview', 'url' => '/core/preview', 'verb' => 'GET'],
-		['name' => 'Preview#getPreview', 'url' => '/core/preview.png', 'verb' => 'GET'],
+		['name' => 'Preview#getPreviewByPath', 'url' => '/core/preview.png', 'verb' => 'GET'],
+		['name' => 'Preview#getPreviewByFileId', 'url' => '/core/preview.fileid', 'verb' => 'GET'],
```
Comment (MorrisJobke) [inline core/routes.php]: `/core/preview.fileid.png` for our friend IE11? 😉 (not tested but just a wild guess)

**Comment by MorrisJobke** [comment]
> @MorrisJobke this is what you had in mind right?

Yes 👍

**Comment by MorrisJobke** [comment]
And the unit tests fail.

**Comment by rullzer** [comment]
And now they pass :)

File: apps/files/js/filelist.js [hunk 2]
```diff
@@ -1856,7 +1857,15 @@
 			urlSpec.x = Math.ceil(urlSpec.x);
 			urlSpec.y = Math.ceil(urlSpec.y);
 			urlSpec.forceIcon = 0;
-			return OC.generateUrl('/core/preview.png?') + $.param(urlSpec);
+
+			if (typeof urlSpec.fileId !== 'undefined') {
+				delete urlSpec.file;
+				return OC.generateUrl('/core/preview.fileid?') + $.param(urlSpec);
```
Comment (nickvergessen) [inline apps/files/js/filelist.js]: can we have a valid file ending please? Helps in some browsers when you right-click save stuff.

Comment (rullzer) [reply apps/files/js/filelist.js]: No we can't. becaus I don't know what type of file we'll server png or jpeg. So whichever one we chose it can lie.

Comment (juliushaertl) [reply apps/files/js/filelist.js]: But that is also the behaviour of the /core/preview.png route, right?

Comment (rullzer) [reply apps/files/js/filelist.js]: Yes. But that is not an argument ;-)

Comment (juliushaertl) [reply apps/files/js/filelist.js]: Ok, just seemed odd to me that we then remove the preview route without the .png extension which would actually be the one that has a more correct :wink:

Comment (juliushaertl) [reply apps/files/js/filelist.js]: https://github.com/nextcloud/server/pull/7902/files#diff-c6f5dabe4eb246c25afe25b39269f751L60

Comment (rullzer) [reply apps/files/js/filelist.js]: yeah that route didn't do anything... was just a leftover. Because we map function to route. so it was overwritten :P

Comment (juliushaertl) [reply apps/files/js/filelist.js]: Ok, fine by me, just seemed a bit inconsistent :see_no_evil:

**Review by juliusknorr — APPROVED** [review]

File: core/routes.php (hunk 1 above)
Comment (rullzer) [reply core/routes.php]: No IE11 works fine if you set the proper content type ;)

**Review by MorrisJobke — APPROVED** [review]

**Pull request merged**